File: Backend/src/routes/schemes.py

from fastapi import APIRouter, HTTPException
from src.aws_db import db_service

router = APIRouter()

# In-memory fallback when DynamoDB is not configured (no API keys required)
_SCHEMES_FALLBACK = [
    {"id": "s1", "name": "PM-Kisan Samman Nidhi", "description": "₹6,000 yearly direct income support for small and marginal farmers.", "status": "Active", "category": "Agriculture"},
    {"id": "s2", "name": "Ayushman Bharat (PM-JAY)", "description": "Free health cover up to ₹5 Lakhs per family per year.", "status": "Active", "category": "Health"},
    {"id": "s3", "name": "PM Awas Yojana (Gramin)", "description": "Housing for all in rural areas.", "status": "Active", "category": "Housing"},
]

@router.get("")
async def get_schemes(q: str = None, category: str = None, state: str = None, lang: str = "en-US"):
    schemes = db_service.get_all_schemes()
    if not schemes:
        schemes = list(_SCHEMES_FALLBACK)

    # Multilingual support: try to return fields in requested language, fallback to English
    def localize(item):
        def pick(field):
            val = item.get(field)
            if isinstance(val, dict):
                # e.g. {"en": "...", "hi": "...", "te": "..."}
                return val.get(lang) or val.get(lang.split("-")[0]) or val.get("en") or next(iter(val.values()), "")
            return val
        return {
            **item,
            "name": pick("name"),
            "description": pick("description"),
            "category": pick("category"),
            "status": pick("status")
        }

    filtered = []
    for s in schemes:
        s_local = localize(s)
        if q and q.lower() not in (s_local.get("name", "") + s_local.get("description", "") + s_local.get("category", "")).lower():
            continue
        if category and category.lower() not in s_local.get("category", "").lower():
            continue
        if state and state.lower() not in s.get("state_availability", "").lower() and "all india" not in s.get("state_availability", "").lower():
            continue
        filtered.append(s_local)
    return filtered
@router.get("/{scheme_id}")
async def get_scheme(scheme_id: str):
    """Return a single scheme by its ID with full detail fields."""
    scheme = db_service.get_scheme_by_id(scheme_id)
    if not scheme:
        raise HTTPException(status_code=404, detail="Scheme not found")
    return scheme



@router.post("/refresh")
async def refresh_schemes():
    """Triggers the scraper to fetch fresh government schemes."""
    from src.services.scraper_service import scraper_service
    success = scraper_service.scrape_schemes()
    if success:
        return {"message": "Government schemes synchronized live!"}
    raise HTTPException(status_code=500, detail="Failed to sync schemes")

@router.get("/bulk")
async def get_schemes_bulk(lang: str = "en-US"):
    """Return all schemes in bulk for offline caching (language-aware)."""
    schemes = db_service.get_all_schemes()
    if not schemes:
        schemes = list(_SCHEMES_FALLBACK)
    def localize(item):
        def pick(field):
            val = item.get(field)
            if isinstance(val, dict):
                return val.get(lang) or val.get(lang.split("-")[0]) or val.get("en") or next(iter(val.values()), "")
            return val
        return {
            **item,
            "name": pick("name"),
            "description": pick("description"),
            "category": pick("category"),
            "status": pick("status")
        }
    return [localize(s) for s in schemes]

File: Backend/src/routes/schemes.ts

import { Router, Request, Response } from 'express';
import { readSchemesJson } from '../utils/readSchemesJson';

const router = Router();

// GET /schemes
router.get('/', async (req: Request, res: Response) => {
  const schemes = readSchemesJson();
  // Return only summary info for list
  const summary = schemes.map(s => ({
    id: s.id,
    name: s.name,
    category: s.category,
    description: s.description,
    benefit: s.benefit,
    status: s.status || 'Active'
  }));
  res.json(summary);
});

// GET /schemes/:scheme_id
router.get('/:scheme_id', async (req: Request, res: Response) => {
  const { scheme_id } = req.params;
  const schemes = readSchemesJson();
  const scheme = schemes.find(s => s.id === scheme_id);
  if (!scheme) {
    return res.status(404).json({ error: 'Scheme not found' });
  }
  res.json(scheme);
});

// GET /schemes/recommend
router.get('/recommend', async (req: Request, res: Response) => {
  // TODO: Replace with real recommendation logic
  res.json([
    { id: 1, name: 'PM Kisan Samman Nidhi', desc: '₹6000/year for eligible farmers', status: 'Active' }
  ]);
});


// Simple in-memory application store for demo
const userApplications: { [userId: string]: { schemeId: string, status: string, applicationData: any }[] } = {};

// POST /schemes/apply
router.post('/apply', async (req: Request, res: Response) => {
  const { userId, schemeId, applicationData } = req.body;
  if (!userId || !schemeId) {
    return res.status(400).json({ error: 'Missing userId or schemeId' });
  }
  if (!userApplications[userId]) userApplications[userId] = [];
  userApplications[userId].push({ schemeId, status: 'Pending', applicationData });
  res.json({ success: true, message: 'Application submitted', status: 'Pending' });
});


// GET /schemes/application-status?userId=xxx&schemeId=yyy
router.get('/application-status', async (req: Request, res: Response) => {
  const { userId, schemeId } = req.query;
  if (!userId || !schemeId) {
    return res.status(400).json({ error: 'Missing userId or schemeId' });
  }
  const apps = userApplications[userId as string] || [];
  const app = apps.find(a => a.schemeId === schemeId);
  if (!app) {
    return res.status(404).json({ error: 'No application found' });
  }
  res.json({ status: app.status, applicationData: app.applicationData });
});

// GET /schemes/applications
router.get('/applications', async (req: Request, res: Response) => {
  // TODO: Fetch all user applications
  res.json([{ id: 1, scheme: 'PM Kisan Samman Nidhi', status: 'Pending' }]);
});

export default router;
